File: TelefQuito_Server/02.CLIWEB/TELFQUITO_SOAP_JAVA_CLIWEB/services/CompraService.js
const SoapClient = require("./soapclient");

class CompraService {
  constructor() {
    this.wsdlUrl = "http://192.168.8.83:8080/TELFQUITO_SOAP_JAVA/WSCompra?wsdl";
    this.namespace = "http://ws.monster.edu.ec/";
  }

  async comprarEfectivo(carrito, codCedula) {
    try {
      console.log("Comprando en efectivo", carrito, codCedula);
      const args1 = {
        carrito: { telefonoCarrito: carrito },
        codcCedula: codCedula,
      };
      console.log("Args", args1);
      const soapClient = new SoapClient(this.wsdlUrl);
      const client = await soapClient.createClient();

      return new Promise((resolve, reject) => {
        const args = {
          carrito: { telefonoCarrito: carrito },
          codcCedula: codCedula,
        };

        client.comprarEfectivo(args, (err, result) => {
          if (err) {
            console.error("SOAP Request Error:", err);
            return reject(err);
          }
          if (result && result.return) {
            resolve(result.return);
          } else {
            console.error("SOAP Response Error: No return value");
            return reject(new Error("No return value from SOAP response"));
          }
        });
      });
    } catch (error) {
      console.error("Compra Efectivo Error:", error);
      throw error;
    }
  }

  async comprarCredito(carrito, cedula, plazoMeses) {
    try {
      const soapClient = new SoapClient(this.wsdlUrl);
      const client = await soapClient.createClient();

      return new Promise((resolve, reject) => {
        const args = {
          carrito: { telefonoCarrito: carrito },
          cedula: cedula,
          plazoMeses: plazoMeses,
        };

        client.comprarCredito(args, (err, result) => {
          if (err) {
            console.error("SOAP Request Error:", err);
            return reject(err);
          }
          if (result && result.return) {
            resolve(result.return);
          } else {
            console.error("SOAP Response Error: No return value");
            return reject(new Error("No return value from SOAP response"));
          }
        });
      });
    } catch (error) {
      console.error("Compra Credito Error:", error);
      throw error;
    }
  }

  async obtenerFactura(cedula) {
    try {
      const soapClient = new SoapClient(this.wsdlUrl);
      const client = await soapClient.createClient();

      return new Promise((resolve, reject) => {
        const args = {
          cedula: cedula,
        };

        client.obtenerFactura(args, (err, result) => {
          if (err) {
            console.error("SOAP Request Error:", err);
            return reject(err);
          }
          if (result && result.return) {
            resolve(result.return);
          } else {
            console.error("SOAP Response Error: No return value");
            return reject(new Error("No return value from SOAP response"));
          }
        });
      });
    } catch (error) {
      console.error("Obtener Factura Error:", error);
      throw error;
    }
  }

  async consultarTablaAmortizacion(cedula) {
    try {
      const soapClient = new SoapClient(this.wsdlUrl);
      const client = await soapClient.createClient();

      return new Promise((resolve, reject) => {
        const args = {
          cedula: cedula,
        };

        client.consultarTablaAmortizacion(args, (err, result) => {
          if (err) {
            console.error("SOAP Request Error:", err);
            return reject(err);
          }
          if (result && result.return) {
            resolve(result.return);
          } else {
            console.error("SOAP Response Error: No return value");
            return reject(new Error("No return value from SOAP response"));
          }
        });
      });
    } catch (error) {
      console.error("Error in consultarTablaAmortizacion:", error);
      throw error;
    }
  }
  async obtenerFacturaEspecifica(cedula, grupoId) {
    try {
      const soapClient = new SoapClient(this.wsdlUrl);
      const client = await soapClient.createClient();

      return new Promise((resolve, reject) => {
        const args = {
          cedula: cedula,
          grupoId: grupoId,
        };

        client.obtenerFacturaEspecifica(args, (err, result) => {
          if (err) {
            console.error("SOAP Request Error:", err);
            return reject(err);
          }
          resolve(result.return);
        });
      });
    } catch (error) {
      console.error("Obtener Factura Especifica Error:", error);
      throw error;
    }
  }
}

module.exports = new CompraService();

File: TelefQuito_Server/02.CLIESC/BANQUITO_SOAP_JAVA_CLIESC/src/main/java/ec/edu/monster/service/CompraService.java
package ec.edu.monster.service;

import ec.edu.monster.ws.Carrito;
import ec.edu.monster.ws.Factura;
import ec.edu.monster.ws.Tabla;
import ec.edu.monster.ws.WSCompra;
import ec.edu.monster.ws.WSCompra_Service;
import java.util.List;

public class CompraService {
    private final WSCompra port;

    public CompraService() {
        // Inicializar el cliente del web service
        WSCompra_Service servicio = new WSCompra_Service();
        this.port = servicio.getWSCompraPort();
    }
    
    /**
     * Realiza una compra en efectivo de un teléfono.
     * @param carrito
     * @param codcCedula Cédula del comprador
     * @return Mensaje de confirmación o error
     */
    public String comprarEfectivo(Carrito carrito, String codcCedula) {
        return port.comprarEfectivo(carrito, codcCedula);
    }

    /**
     * Realiza una compra a crédito de un teléfono.
     * @param carrito
     * @param cedula Cédula del comprador
     * @param plazoMeses Plazo en meses para la compra a crédito
     * @return Mensaje de confirmación o error
     */
    public String comprarCredito(Carrito carrito, String cedula, int plazoMeses) {
        return port.comprarCredito(carrito, cedula, plazoMeses);
    }

    /**
     * Obtiene la lista de facturas de un usuario por su cédula.
     * @param cedula Cédula del usuario
     * @return Lista de objetos Factura
     */
    public List<Factura> obtenerFactura(String cedula) {
        return port.obtenerFactura(cedula);
    }
    
    
    public List<Factura> obtenerFacturaEspecifica(String cedula, int grupoId){
        return port.obtenerFacturaEspecifica(cedula, grupoId);
    }
   

    /**
     * Obtiene la tabla de amortización para un usuario por su cédula.
     * @param cedula Cédula del usuario
     * @return Lista de objetos Tabla
     */
    public List<Tabla> consultarTablaAmortizacion(String cedula) {
        return port.consultarTablaAmortizacion(cedula);
    }
    
}
